x-pack/plugins/alerting/server/task_runner/execution_handler.ts

```diff
@@ -0,0 +1,419 @@
+    this.executionId = executionId;
+    this.ruleLabel = ruleLabel;
+    this.request = request;
+    this.ephemeralActionsToSchedule = taskRunnerContext.maxEphemeralActionsPerRule;
+    this.mutedAlertIdsSet = new Set(rule.mutedInstanceIds);
+  }
+
+  public async run(
+    alerts: Record<string, Alert<State, Context, ActionGroupIds | RecoveryActionGroupId>>,
+    recovered: boolean = false
+  ) {
+    const {
+      CHUNK_SIZE,
+      logger,
+      alertingEventLogger,
+      ruleRunMetricsStore,
+      taskRunnerContext: { actionsConfigMap, actionsPlugin },
+      taskInstance: {
+        params: { spaceId, alertId: ruleId },
+      },
+    } = this;
+
+    const executables = this.generateExecutables({ alerts, actions: this.rule.actions, recovered });
+
+    if (!!executables.length) {
+      const logActions = [];
+      const bulkActions: EnqueueExecutionOptions[] = [];
+      const actionsClient = await this.taskRunnerContext.actionsPlugin.getActionsClientWithRequest(
+        this.request
+      );
+
+      this.ruleRunMetricsStore.incrementNumberOfGeneratedActions(executables.length);
+
+      for (const { action, alert, alertId } of executables) {
+        const { actionTypeId } = action;
+        const actionGroup: ActionGroupIds | RecoveryActionGroupId = recovered
+          ? this.ruleType.recoveryActionGroup.id
+          : alert.getScheduledActionOptions()?.actionGroup!;
+
+        const instanceState = recovered ? {} : alert.getScheduledActionOptions()?.state!;
+
+        if (!recovered) {
+          alert.updateLastScheduledActions(action.group as ActionGroupIds);
+          alert.unscheduleActions();
+        }
+
+        ruleRunMetricsStore.incrementNumberOfGeneratedActionsByConnectorType(actionTypeId);
+
+        if (ruleRunMetricsStore.hasReachedTheExecutableActionsLimit(actionsConfigMap)) {
+          ruleRunMetricsStore.setTriggeredActionsStatusByConnectorType({
+            actionTypeId,
+            status: ActionsCompletion.PARTIAL,
+          });
+          logger.debug(
+            `Rule "${this.rule.id}" skipped scheduling action "${action.id}" because the maximum number of allowed actions has been reached.`
+          );
+          break;
+        }
+
+        if (
+          ruleRunMetricsStore.hasReachedTheExecutableActionsLimitByConnectorType({
+            actionTypeId,
+            actionsConfigMap,
+          })
+        ) {
+          if (!ruleRunMetricsStore.hasConnectorTypeReachedTheLimit(actionTypeId)) {
+            logger.debug(
+              `Rule "${this.rule.id}" skipped scheduling action "${action.id}" because the maximum number of allowed actions for connector type ${actionTypeId} has been reached.`
+            );
+          }
+          ruleRunMetricsStore.setTriggeredActionsStatusByConnectorType({
+            actionTypeId,
+            status: ActionsCompletion.PARTIAL,
+          });
+          continue;
+        }
+
+        if (!this.isActionExecutable(action)) {
```

Comment:
this check can be moved to `generateExecutables` method but then `numberOfGeneratedActions` would be calculated different than it is currently.  How important is this for us?